modules/caddyhttp/reverseproxy/streaming.go

```diff
@@ -209,6 +229,56 @@ func (h Handler) copyBuffer(dst io.Writer, src io.Reader, buf []byte) (int64, er
 	}
 }
 
+// registerConnection holds onto conn so it can be closed in the event
+// of a server shutdown. This is useful because hijacked connections or
+// connections dialed to backends don't close when server is shut down.
+// The caller should call the returned delete() function when the
+// connection is done to remove it from memory.
+func (h *Handler) registerConnection(conn io.ReadWriteCloser, gracefulClose func() error) (delete func()) {
+	h.connections.Store(conn, openConnection{conn, gracefulClose})
+	return func() {
+		h.connections.Delete(conn)
+	}
+}
+
+// writeCloseControl sends a best-effort Close control message to the given
+// WebSocket connection. Thanks to @pascaldekloe who provided inspiration
+// from his simple implementation of this I was able to learn from at:
+// github.com/pascaldekloe/websocket.
+func writeCloseControl(conn io.Writer) error {
+	// https://github.com/pascaldekloe/websocket/blob/32050af67a5d/websocket.go#L119
+
+	var reason string // max 123 bytes (control frame payload limit is 125; status code takes 2)
+	const goingAway uint16 = 1001
+
+	// TODO: we might need to ensure we are the exclusive writer by this point (io.Copy is stopped)?
+	var writeBuf [127]byte
+	const CloseMessage = 8
+	const finalBit = 1 << 7
+	writeBuf[0] = CloseMessage | finalBit
+	writeBuf[1] = byte(len(reason) + 2)
+	binary.BigEndian.PutUint16(writeBuf[2:4], goingAway)
+	copy(writeBuf[4:], reason)
+
+	// simply best-effort, but return error for logging purposes
+	_, err := conn.Write(writeBuf[:4+len(reason)])
+	return err
+}
+
+// isWebsocket returns true if r looks to be an upgrade request for WebSockets.
+// It is a fairly naive check.
+func isWebsocket(r *http.Request) bool {
+	return strings.Contains(strings.ToLower(r.Header.Get("Connection")), "upgrade") &&
+		strings.Contains(strings.ToLower(r.Header.Get("Upgrade")), "websocket")
```

Comment:
Probably doesn't matter but IIRC `Upgrade` needs to be `== "websocket"`, not just a contains. But this is fine.

Reply:
According to MDN, both the Upgrade and Connection headers can have comma-separated list of values:

- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade#syntax
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection#directives

Took my cue from this lib: https://github.com/gorilla/websocket/blob/v1.5.0/server.go#L325